Network retry issue 1602

sbeckeriv opened the pull request:
Dearest Reviewer,

This branch resolves #1602 which relates to retrying network
issues automatically. There is a new utility helper for retrying
any network call.

There is a new config called net.retry value in the .cargo.config
file. The default value is 2. The documentation has also been
updated to reflect the new value.

Thanks
Becker

alexcrichton commented:
Thanks for the PR @sbeckeriv! Some thoughts of mine:

* Could this logic be centralized as much as possible? Currently the retry loop is encoded in two separate places and ideally it'd only be defined in one.
* I'm pretty wary of retrying on *any* error, I've generally had bad experiences with tools that have that kind of behavior. Most of the time an error is impossible to resolve until something changes, it's only pretty rarely that an intermittent error can be retried to fix behavior. Can we whitelist specific errors to retry on?
* I think that this logic needs to be attached directly to the calls to the network themselves, right now it's placed very high up in the chain and far away from where some updates may happen. There's also network updates throughout the rest of cargo (e.g. when publishing or installing) which won't get affected in the locations this is added.
* Can you be sure to add some tests for this? It may be easiest to just mirror the HTTPS tests where some local sever is spun up, connected to, and immediately dropped to generate an error.
* Can you be sure to document any new options in `src/doc/config.md`?

sbeckeriv commented:
Thanks for the reply. 

I will move the net_retry logic in to the utils config object.  For the logic it self I believe I know how to do this. I am going to write a function that takes a closure and handles the logic. 

I will figure out a white list enum too. 

> logic needs to be attached directly to the calls to the network themselves
I think with the closure concept I could push the logic down the stack, however, that would not blanket all network calls.  Do you have some thoughts on how I might go about that? I think most network calls are else where crates.  I could update curl-rust to support a retry option pretty easy. Make it an http setting like proxy and timeout. I read that libcurl does not directly support it. I need to read more around the ssh side. Let me know your thoughts.

alexcrichton commented:
>  Do you have some thoughts on how I might go about that? I think most network calls are else where crates. I could update curl-rust to support a retry option pretty easy.

Nah I can help audit after the infrastructure in place, and then I can also help to ensure that we always push network requests through the same location. I'd also recommend that even if this were an option in curl-rust that we'd still do it in Cargo as we'd likely want to give a UI update as part of it as well anyway.

sbeckeriv commented:
Before I get to far thoughts on these changes:
net retry value from the config with a default value:
https://github.com/rust-lang/cargo/pull/2396/files#diff-4559bc8f75ebf0a4d42be6e4e7fe9eaaR200

Wrapped retry logic:
https://github.com/rust-lang/cargo/pull/2396/files#diff-f9445846d299388561a3972b309c9d9eR1
and an example call:
https://github.com/rust-lang/cargo/pull/2396/files#diff-591901395b44f65a638e1c81531878d2R173

Thanks! It was fun to figure out the with_retry.

alexcrichton commented:
Yeah, that looks great to me!

I think we may want to have another callback for "was this error one we can retry from", but other than that it looks like the right infrastructure to me

sbeckeriv commented:
Hey,
I have updated to the new callback. I updated the docs and added in some comments in the code. I copied a https test and I am checking the stdout. 

For the "was this error one we can retry from"; error handling is new to me. Should I pass it the CargoResult value or the unwrapped Err? What might that callback look like? How do I check its a Err I care about vs any other error? 

Thanks!

alexcrichton commented on src/cargo/core/registry.rs:
```diff
@@ -134,7 +140,7 @@ impl<'cfg> PackageRegistry<'cfg> {
                 if previous.precise() == namespace.precise() {
                     debug!("load/match    {}", namespace);
-                    return Ok(())
+                    return Ok(());
```
Perhaps you have rustfmt on auto-run? Could you back out these changes for now? They tend to clutter up the diff unfortunately :(

alexcrichton commented on src/cargo/core/registry.rs:
```diff
@@ -93,7 +97,8 @@ impl<'cfg> PackageRegistry<'cfg> {
 
         for (_, source) in self.sources.sources_mut() {
-            try!(source.download(package_ids));
+            try!(network::with_retry(&self.config, || source.download(package_ids)));
```
This probably wants to be attached to the literal network calls themselves rather than way up here at the `download` layer. For example a download from everything but the registry doesn't actually need to be retried.

alexcrichton commented on src/cargo/sources/git/source.rs:
```diff
@@ -169,7 +169,10 @@ impl<'cfg> Source for GitSource<'cfg> {
-            let repo = try!(self.remote.checkout(&self.db_path));
+
+            let repo = try!(network::with_retry(&self.config, ||{
+              self.remote.checkout(&self.db_path)
```
Could you attach this to the actual `fetch` done by the git repo?

sbeckeriv replied:
Im going to need some advice here. I started to put a  `cargo_config: &'cfg Config` attribute on `GitRemote`. This lead to a cascade of lifetime markers. After that cleared up I am getting more errors. I think I can reason about the errors but I just wanted to make sure my approach was right. 
Thanks!
Becker


[current list of errors](https://gist.github.com/sbeckeriv/3171ff0b46588c49fd51)

alexcrichton replied:
Ah it should be fine to probably just pass in the Config instead of storing it. Each source already stores a pointer to `Config` somewhere so it should be readily available to pass in

alexcrichton commented:
I think that it may be good to introduce a new trait:

```rust
trait NetworkError {
    pub fn may_be_spurious(&self) -> bool;
}
```

That way you can define the retry function with a `NetworkError` bound on the `E` type coming out. I would only implement `NetworkError` for types like `git2::Error` and `curl::Error` rather than a blanket `CargoError` because that'll help target the errors individually.

bors commented:
:umbrella: The latest upstream changes (presumably #2397) made this pull request unmergeable. Please resolve the merge conflicts.

sbeckeriv commented:
I pushed down the network call in to the checkout command. 

I added the NetworkError trait. I am unsure how to apply the bounds in the retry method. I added it as a trait bound to T but the T is the return Ok value for the cargo result not the error. Any advice or examples of how to do this would be great! 

Thanks again
Becker

alexcrichton commented on src/cargo/util/network.rs:
```diff
@@ -0,0 +1,45 @@
+/// Example:
+/// use util::network;
+/// cargo_result = network.with_retry(&config, || something.download());
+pub fn with_retry<T, F>(config: &Config, mut callback: F) -> CargoResult<T>
```
I think the signature you want for this method is something along the lines of:

```rust
pub fn with_retry<T, E, F>(config: &Config, mut callback: F) -> CargoResult<T>
    where F: FnMut() -> Result<T, E>,
          E: NetworkError + CargoError
```

This would then `match` on the result of `callback` and call `maybe_spurious` to see whether it is indeed a spurious error or not.

alexcrichton commented on src/cargo/util/errors.rs:
```diff
@@ -286,6 +286,16 @@ impl From<Box<CargoError>> for CliError {
+// NetworkError trait
+
+pub trait NetworkError: Error + Send + 'static {
+    fn maybe_spurious(&self) -> bool { true }
```
Yeah this looks along the right lines to me, but you may want to define it as:

```rust
pub trait NetworkError: CargoError {
    fn maybe_spurious(&self) -> bool;
}
```

That way you can elide the extra bounds by just inheriting from `CargoError`, and it also forces the implementations below to check correctly whether they're related to the network before unconditionally retrying.

sbeckeriv replied:
Did not work out of the box. I need some help deciphering what to do.  I pushed the failing code but here is my errors: https://gist.github.com/sbeckeriv/bd2173d07e98b0e599d6

Thanks.
Becker

sbeckeriv replied:
I think I put this on the wrong comment.  I dont think I want to bound the with_retry to the NetworkError trait. I think I want to check in with_retry if the error is NetworkError or not, but all the errors should all be CargoError bound?

alexcrichton commented on src/cargo/util/network.rs:
```diff
@@ -0,0 +1,46 @@
+    let mut result: CargoResult<T> = Err(human("failed to initialize network call"));
+    for try in 0..net_retry + 1 {
+        result = callback();
+        if result.is_err() {
```
Shouldn't this call `maybe_spurious`?

sbeckeriv replied:
What do you see maybe_spurious doing? I am thinking a is_retryable makes more sense. Then it would be `if result.is_err() && result.is_retryable()` which reads like I am thinking the code works. Right now it would return true for the two errors but it could do more for other errors.

alexcrichton commented:
@sbeckeriv yeah the point of the bounds was to highlight the fact that `Box<CargoError>` is opaque and very difficult to see if we need to retry it or not. It basically just means that the retry needs to be pushed farther down in the stack.

sbeckeriv commented:
The bounds make perfect sense now. I pushed the config down to the bottom fetch call. I am still getting "mismatched type" errors on the result I am returning. The generics at this level are beyond me. The error text https://gist.github.com/sbeckeriv/3866bd2ab7241da32b7c  I have tried many combinations. I thought the error came from how I initialized the result value but I removed that. The normal human err is does not impl network error. I then thought it was because the shell command internally could cause an error but I removed the try with no luck. 

Do i need to rebuild the result or implement a from?

Thanks again,
Becker

alexcrichton commented:
No worries! Here's how I might write this function:

```rust
pub fn with_retry<T, E, F>(config: &Config, mut callback: F) -> CargoResult<T>
    where F: FnMut() -> Result<T, E>,
          E: NetworkError
{
    let mut remaining = try!(config.net_retry());
    loop {
        match callback() {
            Ok(ret) => return Ok(ret),
            Err(ref e) if e.maybe_spurious() && remaining > 0 => {
                // print to shell
                remaining -= 1;
            }
            Err(e) => return Err(Box::new(e)),
        }
    }
}
```

Perhaps that can help you out?

sbeckeriv commented:
That does! When I first started I decomposed but I went the other way for reasons I do not remember. I need to update my test but it compiles now. Thanks!

sbeckeriv commented:
I have the retry at the lowest level fetch. I think this covers all the git actions. I am looking in to the interactions with the registry.

sbeckeriv commented:
As I figure out the code I am not sure I want to extern cargo in to crates-io. I am guessing there are separated in to there own project for a reason. Now I would require Config and network to cross back in to creates.   I could just copy the concept but there is no config system for cargo-io helpers. Maybe this is figured out in another pull request? Let me know your thoughts. 

Thanks again!
Becker

alexcrichton commented on src/cargo/util/errors.rs:
```diff
@@ -286,6 +286,20 @@ impl From<Box<CargoError>> for CliError {
+}
+
+impl NetworkError for git2::Error {
+    fn maybe_spurious(&self) -> bool { true }
```
For git errors this may want to at least check the [`class`](http://alexcrichton.com/git2-rs/git2/struct.Error.html#method.class) of the error for `Net`, and perhaps look at the `code` as well to ensure it's not a known non-spurious error.

alexcrichton commented on src/cargo/util/errors.rs:
```diff
@@ -286,6 +286,20 @@ impl From<Box<CargoError>> for CliError {
+    fn maybe_spurious(&self) -> bool { true }
+}
+impl NetworkError for curl::ErrCode {
+    fn maybe_spurious(&self) -> bool { true }
```
We probably want to inspect the error code here a little more to ensure it actually is spurious

sbeckeriv replied:
Im having a heck of a time understand how to use the curl errors.  I believe this should work. 

```curl::ErrCode::COULDNT_CONNECT```

But believing it doesn't make it work. Thoughts on how I can use the names? Should I give up and use the numbers instead?

alexcrichton replied:
Ah using the raw error codes from curl is fine, this should basically just be a whitelist of known bad error codes that can be retried, and we can expand the list over time.

alexcrichton commented:
Looking good! I think though that this may not catch where we download from the registry (using curl), just git so far. API calls with crates.io also look like they may need instrumentation as well, and you can probably just implement `NetworkError` for the error type coming out of the crates-io library

sbeckeriv commented:
I updated the error checking. 

For git2::ErrorCode its not clear which errors relate to network activity. Maybe EOF for a disconnect? In my test I found that not connecting to a host gave an git2::ErrorClass::Os error. I am checking for both Os and Net now. Maybe my test is invalid? 

For curl errors I left a comment to what I believe the number maps to based off of this website https://curl.haxx.se/libcurl/c/libcurl-errors.html 

Thanks again for all the help.
Becker

alexcrichton commented on tests/test_bad_config.rs:
```diff
@@ -227,7 +227,10 @@ test!(bad_git_dependency {
-    .file("src/lib.rs", "");
+    .file("src/lib.rs", "").file(".cargo/config", r#"
+    [net]
+    retry=0
```
In theory these modifications to tests shouldn't be necessary. All of these are indicative of errors that can *never* be resolved, so they shouldn't print out warnings anyway

sbeckeriv replied:
Yes. These are leftover from before the trait. I should be able to remove these.

alexcrichton commented on src/cargo/util/network.rs:
```diff
@@ -0,0 +1,88 @@
+                try!(config.shell()
+                           .say(format!("failed network call. {} \
+                                        tries remaining", remaining),
+                                YELLOW));
```
Could this use `shell().warn(..)`? It also may want to avoid using a period as we tend to avoid that kind of punctuation in Cargo.

alexcrichton commented on src/cargo/util/errors.rs:
```diff
@@ -286,6 +286,37 @@ impl From<Box<CargoError>> for CliError {
+            5  | //COULDNT_RESOLVE_PROXY
+            6  | //COULDNT_RESOLVE_HOST
+            28 | //OPERATION_TIMEDOUT
+            56 // RECV_ERROR
```
Can this link to the curl-sys crate and use the symbolic names for these constants?

sbeckeriv replied:
curl-sys is a dependency of curl but I cant figure out how to access it. Should I just added it as a dependency?

alexcrichton replied:
Yeah adding a dependency is fine

alexcrichton commented on src/cargo/util/errors.rs:
```diff
@@ -286,6 +286,37 @@ impl From<Box<CargoError>> for CliError {
+    fn maybe_spurious(&self) -> bool {
+        match self.class() {
+            git2::ErrorClass::Net |
+            git2::ErrorClass::Os => true,
```
How come `Os` is included here? In theory that's not spurious?

sbeckeriv replied:
This is the class I got for my test when it timed out.  When I tell the test to connect to a server that is not running https://github.com/rust-lang/cargo/pull/2396/files#diff-5a6267705b81a22950ae48ebdce1a553R8 

I thought it would have gave a ```Net``` error but ```Os``` is what it printed.

alexcrichton commented on src/cargo/sources/registry.rs:
```diff
@@ -464,8 +464,9 @@ impl<'cfg> RegistrySource<'cfg> {
-        try!(git::fetch(&repo, &url, refspec).chain_error(|| {
-            internal(format!("failed to fetch `{}`", url))
+
+        try!( git::fetch(&repo, &url, refspec, &self.config).chain_error(|| {
```
looks like an extra leading space snuck in here and the alignment below is a little off

bors commented:
:umbrella: The latest upstream changes (presumably #2486) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #2510) made this pull request unmergeable. Please resolve the merge conflicts.

sbeckeriv commented:
Just an update. I need to merge master but before that I was fighting windows builds output. I have added a new line I have cleaned up newlines. Im not sure what the test is asking for anymore.  It seems to only fail on windows. I need to read over the test code. 

`---- test_cargo_net_config::net_retry_loads_from_config stdout ----
	thread 'test_cargo_net_config::net_retry_loads_from_config' panicked at '
Expected: execs
    but: differences:
  8 -
    + ||
 
 
other output:`

Thanks

alexcrichton commented:
Ah yeah that just means that one more line of output was expected so you just need to end an expectation string with a newline somewhere.

alexcrichton commented:
Ping @sbeckeriv, any update on this?

sbeckeriv commented:
yes I still plan on moving forward with this. I have not had the time in the last few weeks. We can close this until I can get back to it. Also with all my git merging the squash is going to be interesting. I was thinking about doing a patch from this on a fresh branch from master.

alexcrichton commented:
Ah no worries! Was just curious for a status update :)

sbeckeriv commented:
Hello,

Im back on the code. I merged the latest master. I am a little rusty but I am seeing a trait not implemented error https://travis-ci.org/rust-lang/cargo/jobs/127311250 text included below. I use the from_error macro for git2::Error .  I did add a from_error for git2::error::Error but it caught it as a dup. I am on nightly but travis ci has every version failing.  I am going to snoop around but an pointers would be appreciated. 

Thanks again for all the help.
Becker 



`  
 Compiling cargo v0.11.0 (file:///home/travis/build/rust-lang/cargo)
<std macros>:6:1: 6:32 error: the trait `core::convert::From<Box<util::errors::CargoError>>` is not implemented for the type `git2::error::Error` [E0277]
<std macros>:6 $ crate:: convert:: From:: from ( err ) ) } } )
               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/cargo/sources/git/utils.rs:569:9: 571:13 note: in this expansion of try! (defined in <std macros>)
<std macros>:6:1: 6:32 help: run `rustc --explain E0277` to see a detailed explanation
<std macros>:6:1: 6:32 help: the following implementations were found:
<std macros>:6:1: 6:32 help:   <git2::error::Error as core::convert::From<std::ffi::c_str::NulError>>
<std macros>:6:1: 6:32 note: required by `core::convert::From::from`
error: aborting due to previous error
error: Could not compile `cargo`.
`

bors commented:
:umbrella: The latest upstream changes (presumably #2639) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
It looks like `try!` is being called in a function returning a `git2::Error` but the underlying function is returning a `Box<CargoError>` (e.g. a `CarogResult`)?

sbeckeriv commented:
git bisect for the win. Its looking like this change here modified  https://github.com/rust-lang/cargo/commit/537f7185422e8839986ae2ec29a341009b48e418#diff-7c9ac7738c4bef1573b9bbc2945334c8R391

I am wrapping the fetch at the bottom of the call stack here https://github.com/rust-lang/cargo/pull/2396/files#diff-7c9ac7738c4bef1573b9bbc2945334c8R569

I am not sure the correct thing to do here. I think I am going to match on the results of the with_retry and try to return the expected error? Not sure that makes sense. 

Thanks again!

alexcrichton commented:
Sounds plausible to me, yeah

metajack commented:
Servo is starting to hit git errors on a regular basis. We'd really like to see this land. Thanks for working on this @sbeckeriv!

sbeckeriv commented:
@alexcrichton I added whitespace for that windows test but it is still failing. https://github.com/rust-lang/cargo/pull/2396/commits/827d3b6b0c719c53c50ab38d9e9f3bd0060d7082 any thoughts?

I still plan on doing a diff patch instead of squashing this branch. Let me know your thoughts. 
Thanks
Becker

alexcrichton commented:
Ah yeah that looks fine, the appveyor builds are cancelling b/c I'm doing other stuff on other projects (my bad!)

alexcrichton commented on src/cargo/sources/git/utils.rs:
```diff
@@ -565,7 +565,14 @@ pub fn fetch(repo: &git2::Repository, url: &str,
+        }){
+            Ok(_) => Ok(()),
+            Err(e) =>{
+                Err(git2::Error::from_str(&format!("{}", e)))
```
The match shouldn't be needed here, right? I think you can just use `try!`?

sbeckeriv replied:
The try leads me back to this comment https://github.com/rust-lang/cargo/pull/2396#issuecomment-216349535 since the code that wraps this was changed from a CargoResult to a Result. I think the two options are matching and re-packing or changing the signature of  with_authentication back to a CargoResult. Im not sure why the method was changed but this is what made sense to match the new expected return type.  

Is there another way to do this without the match?

alexcrichton replied:
This function returns a `CargoResult`, though, as does `with_retry`, so maybe something changed later on to enable `try!`?

sbeckeriv replied:
with_authentication is a closure that wraps the with_retry. this commit https://github.com/rust-lang/cargo/commit/537f7185422e8839986ae2ec29a341009b48e418#diff-7c9ac7738c4bef1573b9bbc2945334c8R391 changed the return of the with_authentication closure  from CargoResult to Result<T, git2::Error>. There is code for errors like git2::Error to convert to Box<CargoError> but Box<CargoError> does not convert to  Result<git2::Error>. If I am understanding the way errors are setup.

alexcrichton replied:
Aha I see. Hm. So on one hand we need to be a little more proactive here because this will lose error context (`e`'s `Display` implementation only prints itself, not the backtrace of errors). On the other hand this may be difficult to do...

What happens if `with_authentication` requires a closure that returns `CargoResult`?

sbeckeriv replied:
It looks like the test pass for me locally and I can convert back to just calling try! again. I pushed it. I can revert the commit you remember why the CargoResult was turned into a Result.

alexcrichton commented on src/cargo/sources/registry.rs:
```diff
@@ -483,8 +483,9 @@ impl<'cfg> RegistrySource<'cfg> {
-            internal(format!("failed to fetch `{}`", url))
+
+        try!(git::fetch(&repo, &url, refspec, &self.config).chain_error(|| {
+                internal(format!("failed to fetch `{}`", url))
```
Looks like the indentation got a bit off here?

alexcrichton commented on src/cargo/util/errors.rs:
```diff
@@ -1,3 +1,4 @@
+extern crate curl_sys;
```
Could this be added to the crate root?

sbeckeriv replied:
I moved it to src/cargo/lib.rs I think thats the "root". Let me know if that is wrong.

alexcrichton replied:
Ah yeah that's it!

alexcrichton commented on src/cargo/util/errors.rs:
```diff
@@ -286,6 +287,37 @@ impl From<Box<CargoError>> for CliError {
+            curl_sys::CURLcode::CURLE_COULDNT_CONNECT |
+            curl_sys::CURLcode::CURLE_COULDNT_RESOLVE_PROXY |
+            curl_sys::CURLcode::CURLE_COULDNT_RESOLVE_HOST |
+            curl_sys::CURLcode::CURLE_OPERATION_TIMEDOUT |
```
This seems like an operation we shouldn't retry by default?

sbeckeriv replied:
It looks like curl retries (at least once) on CURLE_OPERATION_TIMEDOUT  https://github.com/curl/curl/blob/b9728bca549709a26a5228f1d44f7488dd26811d/src/tool_operate.c#L1449

I do not have any data around which errors people are seeing. It looks like, after a quick search in the curl code, that CURLE_OPERATION_TIMEDOUT covers a lot of different case. After this is merge the list can be changed easy. In this case I believe catching more is better only because if this error code turns out to be bad the work around would be to configure the setting. If you release cargo without this and it turns out that it is the core failure case for curl everyone who is affected need to wait for a new release cycle. 

Let me know if you would still like to remove it.

alexcrichton replied:
Ok, let's leave it in and we can always back it out if weird things happen

alexcrichton commented on src/cargo/util/network.rs:
```diff
@@ -0,0 +1,86 @@
+            Err(ref e) if e.maybe_spurious() && remaining > 0 => {
+                try!(config.shell()
+                           .warn(format!("failed network call {} \
+                                        tries remaining", remaining)));
```
Perhaps this could also print out the error (just for reference)

sbeckeriv replied:
Yes! I agree. I completely removed the "Failed network call" part. It now looks like
" [2/-1] Failed to connect to 127.0.0.1: Connection refused 1 tries remaining"

alexcrichton commented:
Looking good @sbeckeriv, thanks!

alexcrichton commented on src/cargo/util/network.rs:
```diff
@@ -0,0 +1,86 @@
+            Err(ref e) if e.maybe_spurious() && remaining > 0 => {
+                try!(config.shell()
+                           .warn(format!("{} {} tries remaining", e,
+                                         remaining)));
```
Perhaps this could be along the lines of:

```rust
let msg = format!("spurious network error ({} tries remaining): {}", remaining, e);
try!(config.shell().warn(msg));
```

alexcrichton commented:
Awesome! Just a minor tweak to some wording, and then perhaps some squashing? Other than that r=me

sbeckeriv commented:
@alexcrichton I made it one commit. Thanks for working with me and all the help! I learned a lot. 

Becker

alexcrichton commented on src/cargo/util/network.rs:
```diff
@@ -0,0 +1,86 @@
+            Ok(ret) => return Ok(ret),
+            Err(ref e) if e.maybe_spurious() && remaining > 0 => {
+                let msg = format!("spurious network error ({} tries \
+                          remaining): {}", remaining, e);
```
I think the remaining count here is actually `remaining - 1`

sbeckeriv replied:
The retry config is 2 by default.

```
count is 2 
I run once and fail i output 
  2 tries remains 
    subtract 1 count is 1
run a second time 
  1 tries remaining 
    subtract 1 count is 0
retry again
just error. 
```

The retry count is the count of retries not the count of total tries. I can subtract one from the config or update the docs to be explicitly clear it is the count of retries and not the total network tries?

alexcrichton replied:
Ah right yeah that makes sense, nevermind!

alexcrichton commented:
Looks like the Windows CI is failing for non-spurious reasons?

alexcrichton commented:
@bors: r+ 2f04d00ae2b5ddce01b7f5890e5311e287ee198f

bors commented:
:hourglass: Testing commit 2f04d00 with merge e2648ce...

bors commented:
:broken_heart: Test failed - [cargo-win-msvc-64](http://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/438)

sbeckeriv commented:
Sorry I dont have access to a windows box right now. I might tonight.  Does windows not load the .cargo/config file or does it output extra information?

alexcrichton commented:
It should load that, yeah, but I suspect the problem here is that the exact error message is slightly different. It should be possible to just use `with_stderr_contains` perhaps?

alexcrichton commented:
@bors: r+ 7b03532b4f58eea1ccdb397a1b8500fe797df198

bors commented:
:hourglass: Testing commit 7b03532 with merge a4a9491...

bors commented:
:sunny: Test successful - [cargo-cross-linux](http://buildbot.rust-lang.org/builders/cargo-cross-linux/builds/61), [cargo-linux-32](http://buildbot.rust-lang.org/builders/cargo-linux-32/builds/1612), [cargo-linux-64](http://buildbot.rust-lang.org/builders/cargo-linux-64/builds/1601), [cargo-mac-32](http://buildbot.rust-lang.org/builders/cargo-mac-32/builds/1604), [cargo-mac-64](http://buildbot.rust-lang.org/builders/cargo-mac-64/builds/1600), [cargo-win-gnu-32](http://buildbot.rust-lang.org/builders/cargo-win-gnu-32/builds/488), [cargo-win-gnu-64](http://buildbot.rust-lang.org/builders/cargo-win-gnu-64/builds/490), [cargo-win-msvc-32](http://buildbot.rust-lang.org/builders/cargo-win-msvc-32/builds/412), [cargo-win-msvc-64](http://buildbot.rust-lang.org/builders/cargo-win-msvc-64/builds/439)
